source/vsignal/utils.d

```diff
@@ -0,0 +1,148 @@
+then the same is treated as an 'lvalue' and copied.
+
+Params:
+	F = Function to follow.
+	args = a parameter list or an std.meta.AliasSeq.
+
+Returns: An `AliasSeq` of args ready to be correctly forwarded or copied.
+*/
+package template tryForward(alias F, args...)
+{
+	import core.lifetime : forward;
+	import std.meta : AliasSeq;
+
+	// binary search forwardable params
+	template tryForwardImpl(size_t l, size_t r)
+	{
+		alias fwd = forward!(args[l..r]);
+
+		// Safe clause
+		static if (l == r)
+			alias tryForwardImpl = AliasSeq!();
+
+		// If it compiles then F(..., (l .. r), ...) can be forwarded
+		else static if (__traits(compiles, F(args[0..l], fwd, args[r..$])))
+			alias tryForwardImpl = fwd;
+
+		// If we're down to 2 elems just recurse each
+		else static if (r - l == 2)
+			alias tryForwardImpl = AliasSeq!(tryForwardImpl!(l, r - 1), tryForwardImpl!(r - 1, r));
+
+		// If it didn't compile and we have 1 elem left then treat as `lvalue`
+		else static if (r - l == 1)
+			alias tryForwardImpl = args[l..r];
+
+		// If all else fails divide in half
+		else
+			alias tryForwardImpl = AliasSeq!(tryForwardImpl!(l, r / 2), tryForwardImpl!(r / 2, r));
+	}
+
+	alias tryForward = tryForwardImpl!(0, args.length);
+}
+
+@safe pure nothrow @nogc unittest
+{
+	class C
+	{
+		static int foo(int) { return 1; }
+		static int foo(ref int) { return 2; }
+
+		static int foo(int, ref int) { return 1; }
+		static int foo(ref int, int) { return 2; }
+
+		static int foo(ref int, int, const int, const ref int) { return 2; }
+	}
+
+	int foo(Args...)(auto ref Args args)
+	{
+		return C.foo(tryForward!(C.foo, args));
+	}
+
+	int i;
+
+	assert(foo(4) == 1);
+	assert(foo(i) == 2);
+
+	assert(foo(4, i) == 1);
+	assert(foo(i, 4) == 2);
+
+	assert(foo(i, 4, i, 4) == 2);
+
+	static assert(!__traits(compiles, foo(i, i)));
+	static assert(!__traits(compiles, foo(4, 4)));
+}
+
+/**
+Helper for inline imports.
+
+Params:
+	mod = module to import.
+*/
+package template from(string mod)
+{
+	mixin("import from = ", mod, ";");
+}
+
+/**
+Correctly calls a function with the given arguments.
+
+If `F` is not a Callable (e.g lambda infered as void) then `F` is called with
+`args`. If `F` is Callable and is a member function of `args[0]`, then the
+function is called using `args[0]` as the instance with `args[1..$]` as the
+parameters. Otherwise, `F` is called normally. The arguments are attempted to be
+forwarded with `tryForward`.
+
+Params:
+	F = the function to invoke.
+	args = the arguments to call F with.
+
+Examples:
+---
+void method(ref Foo) {}
+
+struct Foo { void method() {} }
+Foo foo;
+
+invoke!(Foo.method)(foo); // correctly invokes method from foo
+invoke!method(foo);       // correctly invokes method
+invoke!(_ => _ + 1)(4);   // correctly invokes the lambda
+---
+
+Returns: The value returned by `F`.
+*/
+package auto ref invoke(alias F, Args...)(auto ref Args args)
+	if (!from!"std.traits".isCallable!F)
```

Comment:
Can't this if be done as a `static if` to prevent having two duplicated function declarations for each case?

Reply:
No, the compiler cannot differentiate both function signatures with `static if`. To read the `static if` it must get into the function's scope, but to do that it must know what function he can get into. These restrictions are attached to the function's signature which helps the compiler differentiate the overloads existent.